### PR TITLE
[clang] Add tests for DRs about inheriting constructors

### PR DESCRIPTION
**Pull request by Endilll**
Covers CWG issues
[1150](https://cplusplus.github.io/CWG/issues/1150.html),
[1487](https://cplusplus.github.io/CWG/issues/1487.html),
[1567](https://cplusplus.github.io/CWG/issues/1567.html),
[1738](https://cplusplus.github.io/CWG/issues/1738.html),
[2273](https://cplusplus.github.io/CWG/issues/2273.html),
[2277](https://cplusplus.github.io/CWG/issues/2277.html),
[2356](https://cplusplus.github.io/CWG/issues/2356.html),
[2504](https://cplusplus.github.io/CWG/issues/2504.html).

On top of the wording in proposed resolutions, [P0136R1](https://wg21.link/p0136r1) "Rewording inheriting constructors (core issue 1941 et al)" is a very relevant paper.

Note that status for 1738 `sup P0136R1` is not officially recognized by CWG, but saying `yes` or `no` seems even more confusing to me. Official resolution is to reject certain code, but Clang is the only implementation that still rejects it to this day: https://godbolt.org/z/b1W8jc1o5. GCC rejected it until 9, now it's accepted: https://godbolt.org/z/of6oh4sdT

### CODE REVIEW

File: clang/test/CXX/drs/dr22xx.cpp [hunk 1]
```diff
@@ -154,6 +154,47 @@ const D &d3(c); // FIXME ill-formed
...
+  B(int);
+  using A::A;
+
+  void f(int); // #dr2277-B-f
+  using A::f;
+};
+
+void g() {
+  B b{0};
+  b.f(0); // FIXME: this is well-formed for the same reason as initialization of 'b' above
```
Comment (cor3ntin) [inline clang/test/CXX/drs/dr22xx.cpp:190]: This in particular looks something we should fix, can you write an issue?

File: clang/test/CXX/drs/dr2504.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+// CHECK: {{.*}}
+
+// SINCE-CXX11-LABEL: define linkonce_odr void @dr2504::B::B()
+// SINCE-CXX11-NOT:     br
+// SINCE-CXX11:         call noundef i32 @dr2504::bar()
+// SINCE-CXX11-NOT:     br
+// SINCE-CXX11:         call void @dr2504::A::A(int)
+// SINCE-CXX11-LABEL: }
```
Comment (cor3ntin) [inline clang/test/CXX/drs/dr2504.cpp:33]: Can you add a fixme and create an issue?

**Comment by cor3ntin** [comment]
This looks good.
I think we should create a meta issue to track our remaining inheriting constructor bugs.
Is there anything in P0136R1 that would not be covered by these core issues?

**Review by cor3ntin — APPROVED** [review]

File: clang/test/CXX/drs/dr2504.cpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+//        inherited from `V`.
+
+// SINCE-CXX11-LABEL: define linkonce_odr void @dr2504::B::B()
+// SINCE-CXX11-NOT:     br
+// SINCE-CXX11:         call noundef i32 @dr2504::bar()
+// SINCE-CXX11-NOT:     br
+// SINCE-CXX11:         call void @dr2504::A::A(int)
+// SINCE-CXX11-LABEL: }
+
+// CHECK: {{.*}}
```
Comment (cor3ntin) [inline clang/test/CXX/drs/dr2504.cpp:37]: Missing return before EOF

**Comment by Endilll** [comment]
To be honest, I'm not keen on the idea of filing bug tracker issues for problems accounted for in other way (DR status, `FIXME`), until we get bug tracker in a good shape, or at least have an understanding how project management side of things is going to look.

> Is there anything in P0136R1 that would not be covered by these core issues?

I'm not sure what do you try to understand by asking this question, but the answer is likely "no".

File: clang/test/CXX/drs/dr2504.cpp (hunk 2 above)
Comment (Endilll) [reply clang/test/CXX/drs/dr2504.cpp]: FIXME added

**Review by shafik — APPROVED** [review]
LGTM

**Comment by cor3ntin** [comment]
Thanks @Endilll :tada:

**Pull request merged**